File: backend/src/DiscretionaryPowers.Domain/Workflow/WorkflowMachine.cs
using DiscretionaryPowers.Domain.Enums;

namespace DiscretionaryPowers.Domain.Workflow;

public static class WorkflowMachine
{
    public static WorkflowState CreateInitialState()
    {
        var stepStatuses = new Dictionary<int, StepStatus>();
        for (var i = 1; i <= 10; i++)
            stepStatuses[i] = StepStatus.NotStarted;

        return new WorkflowState(1, DecisionStatus.Draft, stepStatuses);
    }

    public static (bool Allowed, string? Reason) CanTransition(WorkflowState state, StepTransition transition)
    {
        var stepNumber = transition.StepNumber;
        var action = transition.Action;

        if (stepNumber < 1 || stepNumber > 10)
            return (false, "Step number must be between 1 and 10.");

        if (state.DecisionStatus is DecisionStatus.Published or DecisionStatus.Withdrawn)
            return (false, $"Cannot modify steps when decision is {state.DecisionStatus}.");

        var currentStatus = state.StepStatuses[stepNumber];

        if (action == TransitionAction.Start)
        {
            if (currentStatus != StepStatus.NotStarted)
                return (false, $"Step {stepNumber} is already {currentStatus}.");

            for (var i = 1; i < stepNumber; i++)
            {
                var prevStatus = state.StepStatuses[i];
                if (prevStatus is not (StepStatus.Completed or StepStatus.SkippedWithReason))
                    return (false, $"Step {i} must be completed before starting step {stepNumber}.");
            }
            return (true, null);
        }

        if (action == TransitionAction.Complete)
        {
            if (currentStatus != StepStatus.InProgress)
                return (false, $"Step {stepNumber} must be in progress to complete it. Current status: {currentStatus}.");
            return (true, null);
        }

        if (action == TransitionAction.Skip)
        {
            if (string.IsNullOrWhiteSpace(transition.SkipReason))
                return (false, "A reason must be provided when skipping a step.");

            if (currentStatus == StepStatus.Completed)
                return (false, $"Step {stepNumber} is already completed and cannot be skipped.");

            for (var i = 1; i < stepNumber; i++)
            {
                var prevStatus = state.StepStatuses[i];
                if (prevStatus is not (StepStatus.Completed or StepStatus.SkippedWithReason))
                    return (false, $"Step {i} must be completed before skipping step {stepNumber}.");
            }
            return (true, null);
        }

        return (false, "Unknown action.");
    }

    public static WorkflowState ApplyTransition(WorkflowState state, StepTransition transition)
    {
        var (allowed, reason) = CanTransition(state, transition);
        if (!allowed)
            throw new InvalidOperationException(reason);

        var newStatuses = new Dictionary<int, StepStatus>(state.StepStatuses);
        var currentStep = state.CurrentStep;
        var decisionStatus = state.DecisionStatus;

        switch (transition.Action)
        {
            case TransitionAction.Start:
                newStatuses[transition.StepNumber] = StepStatus.InProgress;
                currentStep = transition.StepNumber;
                if (decisionStatus == DecisionStatus.Draft)
                    decisionStatus = DecisionStatus.InProgress;
                break;

            case TransitionAction.Complete:
                newStatuses[transition.StepNumber] = StepStatus.Completed;
                var next = GetNextStep(new WorkflowState(currentStep, decisionStatus, newStatuses));
                if (next.HasValue)
                    currentStep = next.Value;
                else
                    decisionStatus = DecisionStatus.UnderReview;
                break;

            case TransitionAction.Skip:
                newStatuses[transition.StepNumber] = StepStatus.SkippedWithReason;
                var nextAfterSkip = GetNextStep(new WorkflowState(currentStep, decisionStatus, newStatuses));
                if (nextAfterSkip.HasValue)
                    currentStep = nextAfterSkip.Value;
                else
                    decisionStatus = DecisionStatus.UnderReview;
                break;
        }

        return new WorkflowState(currentStep, decisionStatus, newStatuses);
    }

    public static int? GetNextStep(WorkflowState state)
    {
        for (var i = 1; i <= 10; i++)
        {
            var status = state.StepStatuses[i];
            if (status is StepStatus.NotStarted or StepStatus.InProgress)
                return i;
        }
        return null;
    }

    public static bool CanApprove(WorkflowState state) =>
        state.DecisionStatus == DecisionStatus.UnderReview;

    public static bool CanPublish(WorkflowState state) =>
        state.DecisionStatus == DecisionStatus.Approved;

    public static WorkflowState Approve(WorkflowState state)
    {
        if (!CanApprove(state))
            throw new InvalidOperationException("Decision must be under review to approve.");
        return state with { DecisionStatus = DecisionStatus.Approved };
    }

    public static WorkflowState Publish(WorkflowState state)
    {
        if (!CanPublish(state))
            throw new InvalidOperationException("Decision must be approved before publishing.");
        return state with { DecisionStatus = DecisionStatus.Published };
    }

    public static WorkflowState Withdraw(WorkflowState state)
    {
        if (state.DecisionStatus == DecisionStatus.Published)
            throw new InvalidOperationException("Published decisions cannot be withdrawn.");
        return state with { DecisionStatus = DecisionStatus.Withdrawn };
    }

    public static int GetProgress(WorkflowState state)
    {
        var completed = 0;
        for (var i = 1; i <= 10; i++)
        {
            if (state.StepStatuses[i] is StepStatus.Completed or StepStatus.SkippedWithReason)
                completed++;
        }
        return (int)Math.Round((double)completed / 10 * 100);
    }
}

File: src/modules/decisions/workflow.machine.ts
import { DECISION_STEPS, DECISION_STATUSES } from "@/lib/constants";

export type StepStatus = "not_started" | "in_progress" | "completed" | "skipped_with_reason";
export type DecisionStatus = (typeof DECISION_STATUSES)[keyof typeof DECISION_STATUSES];

export interface WorkflowState {
  currentStep: number;
  decisionStatus: DecisionStatus;
  stepStatuses: Record<number, StepStatus>;
}

export interface StepTransition {
  stepNumber: number;
  action: "start" | "complete" | "skip";
  skipReason?: string;
}

/**
 * 10-step workflow state machine for the BVI discretionary powers framework.
 *
 * Rules:
 * - Steps must be completed in order (1 through 10)
 * - A step cannot be started until the previous step is completed
 * - Steps can be skipped only with a documented reason
 * - The decision moves to "under_review" when all steps are completed
 * - The decision must be explicitly approved and published
 */
export const workflowMachine = {
  /**
   * Initialize a new workflow state for a freshly created decision.
   */
  createInitialState(): WorkflowState {
    const stepStatuses: Record<number, StepStatus> = {};
    for (const step of DECISION_STEPS) {
      stepStatuses[step.number] = "not_started";
    }
    return {
      currentStep: 1,
      decisionStatus: "draft",
      stepStatuses,
    };
  },

  /**
   * Validate whether a step transition is allowed.
   */
  canTransition(state: WorkflowState, transition: StepTransition): { allowed: boolean; reason?: string } {
    const { stepNumber, action } = transition;

    // Validate step number
    if (stepNumber < 1 || stepNumber > 10) {
      return { allowed: false, reason: "Step number must be between 1 and 10." };
    }

    // Cannot modify steps if decision is already published or withdrawn
    if (state.decisionStatus === "published" || state.decisionStatus === "withdrawn") {
      return { allowed: false, reason: `Cannot modify steps when decision is ${state.decisionStatus}.` };
    }

    const currentStatus = state.stepStatuses[stepNumber];

    if (action === "start") {
      if (currentStatus !== "not_started") {
        return { allowed: false, reason: `Step ${stepNumber} is already ${currentStatus}.` };
      }
      // Check that all previous steps are completed or skipped
      for (let i = 1; i < stepNumber; i++) {
        const prevStatus = state.stepStatuses[i];
        if (prevStatus !== "completed" && prevStatus !== "skipped_with_reason") {
          return { allowed: false, reason: `Step ${i} must be completed before starting step ${stepNumber}.` };
        }
      }
      return { allowed: true };
    }

    if (action === "complete") {
      if (currentStatus !== "in_progress") {
        return { allowed: false, reason: `Step ${stepNumber} must be in progress to complete it. Current status: ${currentStatus}.` };
      }
      return { allowed: true };
    }

    if (action === "skip") {
      if (!transition.skipReason?.trim()) {
        return { allowed: false, reason: "A reason must be provided when skipping a step." };
      }
      if (currentStatus === "completed") {
        return { allowed: false, reason: `Step ${stepNumber} is already completed and cannot be skipped.` };
      }
      // Check that all previous steps are completed or skipped
      for (let i = 1; i < stepNumber; i++) {
        const prevStatus = state.stepStatuses[i];
        if (prevStatus !== "completed" && prevStatus !== "skipped_with_reason") {
          return { allowed: false, reason: `Step ${i} must be completed before skipping step ${stepNumber}.` };
        }
      }
      return { allowed: true };
    }

    return { allowed: false, reason: "Unknown action." };
  },

  /**
   * Apply a step transition and return the new state.
   */
  applyTransition(state: WorkflowState, transition: StepTransition): WorkflowState {
    const validation = this.canTransition(state, transition);
    if (!validation.allowed) {
      throw new Error(validation.reason);
    }

    const newState = {
      ...state,
      stepStatuses: { ...state.stepStatuses },
    };

    const { stepNumber, action } = transition;

    if (action === "start") {
      newState.stepStatuses[stepNumber] = "in_progress";
      newState.currentStep = stepNumber;
      if (newState.decisionStatus === "draft") {
        newState.decisionStatus = "in_progress";
      }
    }

    if (action === "complete") {
      newState.stepStatuses[stepNumber] = "completed";
      // Advance currentStep to the next incomplete step
      const nextStep = this.getNextStep(newState);
      if (nextStep) {
        newState.currentStep = nextStep;
      } else {
        // All steps complete — move to under_review
        newState.decisionStatus = "under_review";
      }
    }

    if (action === "skip") {
      newState.stepStatuses[stepNumber] = "skipped_with_reason";
      const nextStep = this.getNextStep(newState);
      if (nextStep) {
        newState.currentStep = nextStep;
      } else {
        newState.decisionStatus = "under_review";
      }
    }

    return newState;
  },

  /**
   * Find the next step that needs action.
   */
  getNextStep(state: WorkflowState): number | null {
    for (let i = 1; i <= 10; i++) {
      const status = state.stepStatuses[i];
      if (status === "not_started" || status === "in_progress") {
        return i;
      }
    }
    return null; // All steps complete
  },

  /**
   * Check if the decision can be approved (all steps done, status is under_review).
   */
  canApprove(state: WorkflowState): boolean {
    return state.decisionStatus === "under_review";
  },

  /**
   * Check if the decision can be published (must be approved first).
   */
  canPublish(state: WorkflowState): boolean {
    return state.decisionStatus === "approved";
  },

  /**
   * Approve a decision.
   */
  approve(state: WorkflowState): WorkflowState {
    if (!this.canApprove(state)) {
      throw new Error("Decision must be under review to approve.");
    }
    return { ...state, decisionStatus: "approved" };
  },

  /**
   * Publish a decision (makes it visible on the public portal).
   */
  publish(state: WorkflowState): WorkflowState {
    if (!this.canPublish(state)) {
      throw new Error("Decision must be approved before publishing.");
    }
    return { ...state, decisionStatus: "published" };
  },

  /**
   * Withdraw a decision.
   */
  withdraw(state: WorkflowState): WorkflowState {
    if (state.decisionStatus === "published") {
      throw new Error("Published decisions cannot be withdrawn.");
    }
    return { ...state, decisionStatus: "withdrawn" };
  },

  /**
   * Get the completion percentage of the workflow.
   */
  getProgress(state: WorkflowState): number {
    let completed = 0;
    for (let i = 1; i <= 10; i++) {
      if (state.stepStatuses[i] === "completed" || state.stepStatuses[i] === "skipped_with_reason") {
        completed++;
      }
    }
    return Math.round((completed / 10) * 100);
  },
};
